### PR TITLE
Mobile chat: Pressing enter on a hardware keyboard should send chat message, shift-enter should add newline

### PR DESCRIPTION
**Pull request by cjb**
@keybase/react-hackers 

Seems to work, needs more thorough testing:

- [x] iPad: with and without keyboard
- [x] iPhone: with and without keyboard
- [x] Android phone: with and without keyboard

### CODE REVIEW

File: shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java [hunk 1]
```diff
@@ -449,4 +452,22 @@ public void setBackgroundColor(DarkModePreference pref) {
       mainWindow.setBackgroundDrawableResource(bgColor);
     });
   }
+
+  @Override
+  public boolean dispatchKeyEvent(KeyEvent event) {
```
Comment (jzila) [inline shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java:457]: Does this not trigger off a virtual keyboard?

Comment (cjb) [reply shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java:457]: @jzila doesn't look like it:

> You should never rely on receiving KeyEvents for any key on a soft input method. In particular, the default software keyboard will never send any key event to any application targetting Jelly Bean or later, and will only send events for some presses of the delete and return keys to applications targetting Ice Cream Sandwich or earlier. 

-- https://developer.android.com/reference/android/view/KeyEvent

Comment (cjb) [reply shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java:457]: (would be super useful if you're able to test! gonna struggle to get a bluetooth keyboard connected to android here.)

Comment (jzila) [reply shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java:457]: Ok I have one, let me build and test

**Comment by jzila** [comment]
I tested this with my bluetooth keyboard and several virtual on-screen keyboards, with the following results:

* The bluetooth keyboard worked correctly every time: it always submitted the chat message.
* Gboard worked correctly every time: <enter> never submitted the message.
* The Samsung on-screen keyboard did not work: instead of adding a newline, it actually submitted the message.

I added two changes:
* The first tracks the bluetooth keyboard state. If one is connected, then we dispatch <enter> keyup events to the JS runtime. Otherwise we do not. With this change, the Samsung keyboard worked correctly while the bluetooth keyboard was not connected. But when it was, it did the wrong thing again.
* We track `isOpen` in the JS runtime and only handle <enter> as a submit if the on-screen keyboard is _not_ open. If it is open, we inject a newline. That seems like the right compromise, since as long as you're typing on a bluetooth keyboard, the on-screen keyboard should disappear.

**Comment by cjb** [comment]
@jzila LGTM, thanks!

**Pull request merged**